openwisp_monitoring/views.py

```diff
@@ -64,6 +64,13 @@ def get(self, request, *args, **kwargs):
         # if custom dates are provided then validate custom dates
         if start_date and end_date:
             self._validate_custom_date(start_date, end_date, timezone)
+            # if valid custom dates then calculate custom days
+            time = '1d'
+            end = datetime.strptime(end_date, '%Y-%m-%d %H:%M:%S')
+            start = datetime.strptime(start_date, '%Y-%m-%d %H:%M:%S')
+            custom_days = (end - start).days
+            if custom_days:
+                time = f'{custom_days}d'
```

Comment:
I don't understand: why is this change needed?

Reply:
Because we removed `time` from the query parameter, we must calculate the _`days diff (end - start)`_ after validating the custom date strings in the views in order to update [`custom_group_map`](https://github.com/openwisp/openwisp-monitoring/blob/ec248d1cee0cff8d81714d7c9d50adbab8f858f3/openwisp_monitoring/monitoring/base/models.py#L548-L577) accordingly.

Ref : https://github.com/openwisp/openwisp-monitoring/pull/454#discussion_r1020930821

Reply:
ok thanks for the refersh.

I see that we are parsing the dates twice, which is bad.
We should modify the `_validate_custom_date` so that it returns the dates we need here, I would also rewrite this code as follows:

```python
start_datetime, end_datetime = self._validate_custom_date(start_date, end_date, timezone)
# if valid custom dates then calculate custom days
time = '1d'
custom_days = (start_datetime - end_datetime).days
if custom_days:
    time = f'{custom_days}d'
```

Reply:
thanks! @nemesisdesign